src/main/java/org/opentripplanner/raptor/rangeraptor/TimeBasedRoutingStrategy.java

```diff
@@ -0,0 +1,249 @@
+package org.opentripplanner.raptor.rangeraptor;
+
+import java.util.function.IntConsumer;
+import javax.annotation.Nonnull;
+import org.opentripplanner.raptor.rangeraptor.internalapi.RoundProvider;
+import org.opentripplanner.raptor.rangeraptor.internalapi.RoutingStrategy;
+import org.opentripplanner.raptor.rangeraptor.internalapi.SlackProvider;
+import org.opentripplanner.raptor.rangeraptor.internalapi.WorkerLifeCycle;
+import org.opentripplanner.raptor.rangeraptor.transit.TransitCalculator;
+import org.opentripplanner.raptor.spi.RaptorAccessEgress;
+import org.opentripplanner.raptor.spi.RaptorConstrainedTripScheduleBoardingSearch;
+import org.opentripplanner.raptor.spi.RaptorTimeTable;
+import org.opentripplanner.raptor.spi.RaptorTripSchedule;
+import org.opentripplanner.raptor.spi.RaptorTripScheduleBoardOrAlightEvent;
+import org.opentripplanner.raptor.spi.RaptorTripScheduleSearch;
+import org.opentripplanner.raptor.spi.TransitArrival;
+import org.opentripplanner.routing.algorithm.raptoradapter.transit.request.TripScheduleBoardSearch;
+
+/**
+ * This class contains code which is shared by all time-dependent routing strategies. It also
+ * defines abstract methods, which the individual strategies must implement.
+ */
+public abstract class TimeBasedRoutingStrategy<T extends RaptorTripSchedule>
```

Comment:
Where should this package lie, and what would be a good name?

Reply:
I don't have a strong preference but these are called `worker` in other places, the naming should be consistent.

I'd say the "time dependent" aspect would need to be reflected in the name.

Reply:
I've been thinking about this and I think the involved classes and package name are a bit confusing, if I recap we have:

`Worker `=> Implementations: `RangeRaptorWorker`, `HeuristicSearch`
`HeuristicSearch `=> is a wrapper around a `RangeRaptorWorker ` and an `Heuristics`

`RangeRaptorWorker `=> do range or single iteration raptor with strategy through `RoutingStrategy` transitWorker

Existing strategies:
- `MultiCriteriaRoutingStrategy`
- `ArrivalTimeRoutingStrategy`
- `MinTravelDurationRoutingStrategy`

In their current state only `MultiCriteriaRoutingStrategy` considers the wait time, constraint transfers and do create tripSearch while `ArrivalTimeRoutingStrategy` and `MinTravelDurationRoutingStrategy` do not.

These strategy usually have a `WorkerState` which is either a `McRangeRaptorWorkerState` for multi criteria or a `StdWorkerState` for single criteria strategy.

The state contains various implementations such as `BestTimes` `StopArrivalsState` `McStopArrivals` and some logic to limit search exploration (like destination pruning), update stop reached last round, there's some 'duplication' in these classes, for instance to handle stops reached last round, storing the best values, etc. also when there's more than one criteria involved, each criteria is usually handled in a duplicated way: i.e. `int transfers[]`, `int times[]` and `int costs[]`.

Would it be cleaner to have something like that:

- `RaptorWorker`
  - Role: Range Raptor core algorithm, iterations, transit, transfer, etc.
Implementations: `RangeRaptorWorker` (potentially `SingleIterationRaptorWorker` but I don't think there's a need for it)

- `TripSearchStrategy` or `BoardingStrategy`
  - Role: Selecting proper 'trip-search' strategy, i.e. this new strategy and the basic implementation proposed in `HeuristicRoutingStrategy`
Implementations: `NoWait`, `TimeDependent`

- `RoutingProfile`/`RoutingStrategy` or `RoutingStateStrategy`
  - Role: Selecting criteria/pareto function or algorithm used for updating the state
Implementations: `MultiCriteria`, `RoundOnly`, `ArrivalTime`, `MinDuration`, `DurationAndCost`, etc..
Generalisations possibles: `SingleCriteria<Criteria>`, `MultiCriteria<2,n>` 

- `HeuristicsStrategy`
  - Role: Prevent state update based on various heuristics approaches
Implementations: `MaxRoundHeuristics`, `DestinationPruningHeuristics`, `EuclidianDistanceHeuristics`, `RoutePruningHeuristics` (c.f. long distance discussions)

If we can split it this way instead of the `RoutingStrategy` (state, criterias and heuristics logic) that inherits from the `TripSearchStrategy` (trip search, boarding, transfer constraints) it would be more flexible as these concepts are orthogonal and not necessarily linked.

Also potentially the `TerminationStrategy` or `HeuristicStrategy` could extract the logic for destination pruning or other heuristics check.

It seems that this would allow a larger 'combination' of searches concepts while keeping stricter roles/purpose to each components. The building of the search profile could look similar to this: 
```java
RangeRaptorWorker
    .of(ONE_ITERATION)
    .withBoardingStrategy(new NoWaitStrategy())
    .withRoutingStateStrategy(new ArrivalTimeStrategy())
    .withHeuristics(new MaxRoundHeuristics(bestRound...))
    .build();
```
Which would be easier to understand in my view than the current code in Std and McRaptorConfig classes. And would help with the goal of having several heuristics pass for several criterias.